### ISSUE TITLE
'ForgeParams4bit' object has no attribute 'module'

### ISSUE THREAD

**Issue by caustiq** [comfyanonymous/ComfyUI_bitsandbytes_NF4#12]
```
Error occurred when executing KSampler:

'ForgeParams4bit' object has no attribute 'module'

File "/home/user/Ai/ComfyUI/execution.py", line 152, in recursive_execute
output_data, output_ui = get_output_data(obj, input_data_all)
File "/home/user/Ai/ComfyUI/execution.py", line 82, in get_output_data
return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
File "/home/user/Ai/ComfyUI/execution.py", line 75, in map_node_over_list
results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
File "/home/user/Ai/ComfyUI/nodes.py", line 1382, in sample
return common_ksampler(model, seed, steps, cfg, sampler_name, scheduler, positive, negative, latent_image, denoise=denoise)
File "/home/user/Ai/ComfyUI/nodes.py", line 1352, in common_ksampler
samples = comfy.sample.sample(model, noise, steps, cfg, sampler_name, scheduler, positive, negative, latent_image,
File "/home/user/Ai/ComfyUI/comfy/sample.py", line 43, in sample
samples = sampler.sample(noise, positive, negative, cfg=cfg, latent_image=latent_image, start_step=start_step, last_step=last_step, force_full_denoise=force_full_denoise, denoise_mask=noise_mask, sigmas=sigmas, callback=callback, disable_pbar=disable_pbar, seed=seed)
File "/home/user/Ai/ComfyUI/comfy/samplers.py", line 829, in sample
return sample(self.model, noise, positive, negative, cfg, self.device, sampler, sigmas, self.model_options, latent_image=latent_image, denoise_mask=denoise_mask, callback=callback, disable_pbar=disable_pbar, seed=seed)
File "/home/user/Ai/ComfyUI/comfy/samplers.py", line 729, in sample
return cfg_guider.sample(noise, latent_image, sampler, sigmas, denoise_mask, callback, disable_pbar, seed)
File "/home/user/Ai/ComfyUI/comfy/samplers.py", line 706, in sample
self.inner_model, self.conds, self.loaded_models = comfy.sampler_helpers.prepare_sampling(self.model_patcher, noise.shape, self.conds)
File "/home/user/Ai/ComfyUI/comfy/sampler_helpers.py", line 66, in prepare_sampling
comfy.model_management.load_models_gpu([model] + models, memory_required=memory_required, minimum_memory_required=minimum_memory_required)
File "/home/user/Ai/ComfyUI/comfy/model_management.py", line 526, in load_models_gpu
cur_loaded_model = loaded_model.model_load(lowvram_model_memory, force_patch_weights=force_patch_weights)
File "/home/user/Ai/ComfyUI/comfy/model_management.py", line 323, in model_load
self.model.unpatch_model(self.model.offload_device)
File "/home/user/Ai/ComfyUI/comfy/model_patcher.py", line 614, in unpatch_model
self.model.to(device_to)
File "/home/user/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1340, in to
return self._apply(convert)
File "/home/user/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 900, in _apply
module._apply(fn)
File "/home/user/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 900, in _apply
module._apply(fn)
File "/home/user/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 927, in _apply
param_applied = fn(param)
File "/home/user/.local/lib/python3.10/site-packages/torch/nn/modules/module.py", line 1326, in convert
return t.to(
File "/home/user/Ai/ComfyUI/custom_nodes/ComfyUI_bitsandbytes_NF4/__init__.py", line 66, in to
module=self.module
```
using bitsandbytes 'multi-backend-refactor' branch that supports AMD GPU
pytorch version: 2.5.0.dev20240811+rocm6.1
rocm 6.2 for AMD 6900XT

**Comment by fgdfgfthgr-fox** [comment]
Same issue here. I am also using an AMD gpu (Radeon VII) with multi-backend-refactor bitsandbytes.

**Comment by CambridgeComputing** [comment]
I ran into the same error and upgrading bitandbytes fixed it for me:
`pip install bitsandbytes -U`

**Comment by caustiq** [comment]
> I ran into the same error and upgrading bitandbytes fixed it for me: `pip install bitsandbytes -U`
nah, this issue is AMD specific so we need to build manually from a different branch (mentioned earlier). That command would just install the version without AMD support.

> I'm about to try rocm 6.1 instead of 6.2
I tried rocm 6.1 with pytorch 2.4.0 stable and ran into the same issue in the OP.

**Comment by CambridgeComputing** [comment]
> > I ran into the same error and upgrading bitandbytes fixed it for me: `pip install bitsandbytes -U`
> 
> nah, this issue is AMD specific so we need to build manually from a different branch (mentioned earlier). That command would just install the version without AMD support.
> 
> > I'm about to try rocm 6.1 instead of 6.2
> 
> I tried rocm 6.1 with pytorch 2.4.0 stable and ran into the same issue in the OP.

Apologies, I didn't see the ROCm part of the original report, you are correct.

**Comment by caustiq** [comment]
I'm also getting #15 on startup and since it's related to 'types' and zope.interface I wonder if it's the root cause of this issue?

**Comment by fgdfgfthgr-fox** [comment]
> @fgdfgfthgr-fox which version of pytorch and rocm are you using? I'm about to try rocm 6.1 instead of 6.2

I honestly don't know anymore because I just tried to upgrade rocm and now my computer won't boost with graphic card enabled and I have to solve that first.
I guess that's what you get for using rocm.

**Comment by fgdfgfthgr-fox** [comment]
I suspect it's simply because the multi-backend-refactor branch of bitsandbytes isn't up to date yet, all we can do is wait, or comfyanonymous could use another implementation.

**Comment by initialxy** [comment]
I can confirm that if you merge `main` onto `multi-backend-refactor`, resolve a couple of merge conflicts, which are pretty trivial. Then build and install from there, this error goes away and it will work. I have a RX 7900 XTX and rocm 6.1.

**Comment by gabriel-filincowsky** [comment]
Note: I am using an AMD CPU and a Nvidia GPU. Maybe the problem is not with the AMD GPU but with the AMD version of TORCH or some other resource.


I already tried to:
- Update ComfyUI and dependencies
- Update bitsandbytes


Hardware:
CPU: AMD Ryzen 9 5900HX with Radeon Graphics - Arch: AMD64 - OS: Windows 10
GPU: NVIDIA GeForce RTX 3080 Laptop GPU
NVIDIA Driver: 560.81
Total VRAM 16384 MB, total RAM 32175 MB
pytorch version: 2.5.0.dev20240812+cu121
Set vram state to: NORMAL_VRAM
Device: cuda:0 NVIDIA GeForce RTX 3080 Laptop GPU : cudaMallocAsync


```error
!!! Exception during processing!!! 'ForgeParams4bit' object has no attribute 'module'
Traceback (most recent call last):
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\comfy\model_management.py", line 321, in model_load
    self.real_model = self.model.patch_model(device_to=patch_model_to, patch_weights=load_weights)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\comfy\model_patcher.py", line 352, in patch_model
    self.model.to(device_to)
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\python_embeded\Lib\site-packages\torch\nn\modules\module.py", line 1340, in to
    return self._apply(convert)
           ^^^^^^^^^^^^^^^^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\python_embeded\Lib\site-packages\torch\nn\modules\module.py", line 900, in _apply
    module._apply(fn)
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\python_embeded\Lib\site-packages\torch\nn\modules\module.py", line 900, in _apply
    module._apply(fn)
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\python_embeded\Lib\site-packages\torch\nn\modules\module.py", line 927, in _apply
    param_applied = fn(param)
                    ^^^^^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\python_embeded\Lib\site-packages\torch\nn\modules\module.py", line 1326, in convert
    return t.to(
           ^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\custom_nodes\ComfyUI_bitsandbytes_NF4\__init__.py", line 66, in to
    module=self.module
           ^^^^^^^^^^^
AttributeError: 'ForgeParams4bit' object has no attribute 'module'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\execution.py", line 152, in recursive_execute
    output_data, output_ui = get_output_data(obj, input_data_all)
                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\execution.py", line 82, in get_output_data
    return_values = map_node_over_list(obj, input_data_all, obj.FUNCTION, allow_interrupt=True)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\execution.py", line 75, in map_node_over_list
    results.append(getattr(obj, func)(**slice_dict(input_data_all, i)))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\custom_nodes\ComfyUI_bitsandbytes_NF4\__init__.py", line 178, in load_checkpoint
    out = comfy.sd.load_checkpoint_guess_config(ckpt_path, output_vae=True, output_clip=True, embedding_directory=folder_paths.get_folder_paths("embeddings"), model_options={"custom_operations": OPS})
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\comfy\sd.py", line 511, in load_checkpoint_guess_config
    out = load_state_dict_guess_config(sd, output_vae, output_clip, output_clipvision, embedding_directory, output_model, model_options)
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\comfy\sd.py", line 588, in load_state_dict_guess_config
    model_management.load_model_gpu(model_patcher)
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\comfy\model_management.py", line 540, in load_model_gpu
    return load_models_gpu([model])
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\comfy\model_management.py", line 526, in load_models_gpu
    cur_loaded_model = loaded_model.model_load(lowvram_model_memory, force_patch_weights=force_patch_weights)
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\comfy\model_management.py", line 323, in model_load
    self.model.unpatch_model(self.model.offload_device)
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\comfy\model_patcher.py", line 618, in unpatch_model
    self.model.to(device_to)
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\python_embeded\Lib\site-packages\torch\nn\modules\module.py", line 1340, in to
    return self._apply(convert)
           ^^^^^^^^^^^^^^^^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\python_embeded\Lib\site-packages\torch\nn\modules\module.py", line 900, in _apply
    module._apply(fn)
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\python_embeded\Lib\site-packages\torch\nn\modules\module.py", line 900, in _apply
    module._apply(fn)
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\python_embeded\Lib\site-packages\torch\nn\modules\module.py", line 927, in _apply
    param_applied = fn(param)
                    ^^^^^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\python_embeded\Lib\site-packages\torch\nn\modules\module.py", line 1326, in convert
    return t.to(
           ^^^^^
  File "G:\_SD\Swarm\StableSwarmUI\dlbackend\comfy\ComfyUI\custom_nodes\ComfyUI_bitsandbytes_NF4\__init__.py", line 66, in to
    module=self.module
           ^^^^^^^^^^^
AttributeError: 'ForgeParams4bit' object has no attribute 'module'
```

**Comment by initialxy** [comment]
I can add that I found it needs rocm 6.0 or greater otherwise it will just crash at bitsandbytes when you validate it with `python -m bitsandbytes`

**Comment by gabriel-filincowsky** [comment]
Hi @initialxy,

I've spent considerable time trying to implement your suggestion with the ComfyUI_bitsandbytes_NF4 repository, but I ran into a roadblock. It seems there is no multi-backend-refactor branch in that repository—only master is available. You are probably talking about something else. Sorry, I'm a newbie here 🥺.

Could you please clarify if you're referring to a different repository or perhaps a fork? I would greatly appreciate any additional details to help me move forward.

Thanks for your guidance!

**Comment by hartmark** [comment]
> I can confirm that if you merge `main` onto `multi-backend-refactor`, resolve a couple of merge conflicts, which are pretty trivial. Then build and install from there, this error goes away and it will work. I have a RX 7900 XTX and rocm 6.1 on Arch Linux.

Care to share a fork with a branch with `main` and `multi-backend-refactor` merged?

**Comment by gabriel-filincowsky** [comment]
I'm trying to brut force to learn how all this work. This is what I talked with gpt-4o. Please let me know if any of this makes sense.

```me
I'm not sure, but I think that the ComfyUI_bitsandbytes_NF4 repository is only a small adaptation for bitsandbytes to run in the comfyui (like a plug-in).

I think that what that user was referring to is the original repository for bitsandbytes. When looking for it on GitHub I found the following repository.
https://github.com/bitsandbytes-foundation/bitsandbytes/tree/multi-backend-refactor

You can notice that in this repository we have the mentioned branch. Now, the question is how does this relates to comfyui? Would it be the case that bitsandbytes is installed in another directory as a requirement, maybe in the pythonembeded folder? Or could be the case the ComfyUI_bitsandbytes_NF4 repository was created using the main from https://github.com/bitsandbytes-foundation/bitsandbyte and would require to create a fork of ComfyUI_bitsandbytes_NF4 using the multi-backend-refactor branch????

What do you think about this? Can you please offer a better insight and direction?

I am asking it to you. I'm not trying to review the previous post text.
```

```gpt-4o
It sounds like you've uncovered an important piece of the puzzle. The `ComfyUI_bitsandbytes_NF4` repository is indeed likely a plugin or adaptation to get `bitsandbytes` working specifically within ComfyUI. The branch `multi-backend-refactor` you mentioned from the original `bitsandbytes` repository is indeed a major branch in the official repository for `bitsandbytes` (https://github.com/bitsandbytes-foundation/bitsandbytes).

Given this, here’s what could be happening:

1. **ComfyUI_bitsandbytes_NF4 as an Adapter**: The `ComfyUI_bitsandbytes_NF4` repository is a wrapper or plugin to make the `bitsandbytes` library compatible with ComfyUI. It likely pulls in `bitsandbytes` as a dependency.

2. **How It Relates to ComfyUI**:
    - **`bitsandbytes` in the Python Environment**: The actual `bitsandbytes` library is likely installed as a dependency within the `python_embeded` folder or another location where ComfyUI's Python environment manages packages.
    - **Updating or Modifying `bitsandbytes`**: To work with the `multi-backend-refactor` branch, you might need to manually clone the `bitsandbytes` repository from the official source (https://github.com/bitsandbytes-foundation/bitsandbytes/tree/multi-backend-refactor), build it, and replace the existing `bitsandbytes` installation in your environment.

3. **Potential Approaches**:
    - **Direct Update**: Clone the `multi-backend-refactor` branch from the official `bitsandbytes` repository, build it, and replace the installed `bitsandbytes` in your environment (`python_embeded`).
    - **Fork and Modify**: If the `ComfyUI_bitsandbytes_NF4` repository is tightly coupled with a specific branch from the original `bitsandbytes`, you may consider forking `ComfyUI_bitsandbytes_NF4`, updating it to work with `multi-backend-refactor`, and then integrating it back into ComfyUI.

### Steps to Proceed:

1. **Clone the `multi-backend-refactor` Branch**:
   ```bash
   git clone -b multi-backend-refactor https://github.com/bitsandbytes-foundation/bitsandbytes.git
   cd bitsandbytes
   ```
   Build it according to the instructions provided in the repository.

2. **Replace Existing Installation**:
   Identify where `bitsandbytes` is currently installed (likely in `python_embeded/lib/site-packages` or similar) and replace the existing files with your new build.

3. **Test**: Run the setup and see if the issue is resolved within ComfyUI.

This approach should help integrate the updates from the `multi-backend-refactor` branch into your ComfyUI environment. If you encounter any specific issues during this process, further adjustments might be needed.
```

**Comment by initialxy** [comment]
@gabriel-filincowsky o I was referring to manually replacing the bitsandbytes dependency from its original repo https://github.com/bitsandbytes-foundation/bitsandbytes/tree/multi-backend-refactor

Even though I mentioned that merging `main` to `multi-backend-refactor` is pretty trivial, I don't feel comfortable submitting a PR since I'm not if maintainers prefer to merge `multi-backend-refactor` to `main` instead. I'd recommend just waiting for the official bitsandbytes repo to merge and release a new version.

In the meantime, if you really want to try, I forked it and performed a merge (though I will delete this repo when upstream is updated). https://github.com/initialxy/bitsandbytes/tree/multi-backend-refactor

    # activate your local python venv
    source venv/bin/activate
    # uninstall the current installation of bitsandbytes
    pip uninstall bitsandbytes
    # clone the above
    git clone https://github.com/initialxy/bitsandbytes.git
    cd bitsandbytes
    git checkout multi-backend-refactor
    # build it, assuming you have all the build tools. I just followed AUR's PKGBUILD script https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=python-bitsandbytes-rocm-git
    cmake -DCOMPUTE_BACKEND=hip -S .
  make
    pip install build installer
    python -m build --wheel --no-isolation
    python -m installer --destdir="$pkgdir" $pkgname/dist/*.whl
    # check that it installed successfully
    python -m bitsandbytes

and that worked for me. Again, I'd recommend just wait for the official repo to merge. Things are moving really fast right now. Also, reading your prior comment, you mentioned that you have an AMD and Nvidia GPUs, so I'm not super sure if this will work. This rocm build is only meant for AMD GPU.

**Comment by hartmark** [comment]
Thanks for the fork, I'll check it out when I get the time. 

Regarding mixing amd and Nvidia. I don't know if it's supported because for Nvidia you use DCOMPUTE_BACKEND=cuda

**Comment by gabriel-filincowsky** [comment]
@initialxy, Thank you! I really appreciate you taking the time to reply.

**Comment by Jeremy8776** [comment]
So we all still waiting on a fix?

**Comment by hartmark** [comment]
Any news on this one?

**Comment by Jeremy8776** [comment]
> Any news on this one?

Ive been using forge for now, also a lot quicker

**Comment by hartmark** [comment]
> > Any news on this one?
> 
> Ive been using forge for now, also a lot quicker

Interesting, do you have any details on how much faster? and what is your hardware specs?

**Comment by Jeremy8776** [comment]
Specs: 3090FE [24gb], 32gb RAM, Ryzen 5 3600x

Speeds on comfy can take me between 1min  - 2 mins a gen. Not using nf4 model due to the above bug.

Speeds on Forge can take me 30s-40s, 2mins inc hiresfix and 2xupscale . Using NF4 model

Both with 35 steps, Euler, Normal

**Comment by hartmark** [comment]
Thanks for details, I'm going to try out forge a bit and see how it will work. 

I have AMD 7800 xt 16GB
Ryzen 9 5900 12c
32GB ram

I'm having issues with lockups due to vram not getting swapped out correctly.

I got flux schnell working with GGUF and have 30-50s to generate 1024x1024

https://github.com/city96/ComfyUI-GGUF

I have tried getting controlnet working but i get consistent lockups with it
https://www.reddit.com/r/StableDiffusion/comments/1euz2a9/comment/lio2fte/?utm_source=share&utm_medium=web3x&utm_name=web3xcss&utm_term=1&utm_content=share_button

**Comment by hartmark** [comment]
My random lockups was due to this issue.
https://github.com/ROCm/ROCm/issues/3580

TLDR:
The issue is in the kernel driver and most issues are resolved now in later kernel versions, with this part still lingering
https://github.com/comfyanonymous/ComfyUI/issues/5341

**Comment by jnolck** [comment]
> I can confirm that if you merge `main` onto `multi-backend-refactor`, resolve a couple of merge conflicts, which are pretty trivial. Then build and install from there, this error goes away and it will work. I have a RX 7900 XTX and rocm 6.1 on Arch Linux.

Sigh, that was a pain in the butt for me, but I got it to work! I'm disappointed though, I thought I'd get a good speedup but I didn't. The vram on this model with this node goes down to 60% max and the quality of the output is very good. It only shaved off a few seconds off my generation speed down to 35seconds for generation compared to 39 with the fp8 model and nodes.

**Comment by hartmark** [comment]
> Sigh, that was a pain in the butt for me, but I got it to work! I'm disappointed though, I thought I'd get a good speedup but I didn't. The vram on this model with this node goes down to 60% max and the quality of the output is very good. It only shaved off a few seconds off my generation speed down to 35seconds for generation compared to 39 with the fp8 model and nodes.

Cool, keep me updated if it helps the performance. Do you have any branch with the merge so others can try it out?

**Comment by jnolck** [comment]
> > Sigh, that was a pain in the butt for me, but I got it to work! I'm disappointed though, I thought I'd get a good speedup but I didn't. The vram on this model with this node goes down to 60% max and the quality of the output is very good. It only shaved off a few seconds off my generation speed down to 35seconds for generation compared to 39 with the fp8 model and nodes.
> 
> Cool, keep me updated if it helps the performance. Do you have any branch with the merge so others can try it out?

No. I have no idea how to do that. I did edit a lot of files too. Some nasty hacky stuff in there.
